cdc/owner/scheduler.go

```diff
@@ -0,0 +1,301 @@
+	s.captureWorkloadDeltas = make(map[model.CaptureID]int)
+
+	// We do NOT want to touch these tables because they are being deleted.
+	// We wait for the deletion(s) to finish before redispatching.
+	pendingList := s.cleanUpOperations()
+	pendingSet := make(map[model.TableID]struct{})
+	for _, tableID := range pendingList {
+		pendingSet[tableID] = struct{}{}
+	}
+
+	tableToCaptureMap := s.ownerState.GetTableToCaptureMap(s.cfID)
+
+	// handle adding table
+	for tableID, task := range tables {
+		if _, ok := pendingSet[tableID]; ok {
+			// Table has a pending deletion. Skip.
+			continue
+		}
+
+		if _, ok := tableToCaptureMap[tableID]; !ok {
+			// Table is not assigned to a capture.
+			target := s.getMinWorkloadCapture()
+			if target == "" {
+				log.Warn("no capture is active")
+				break
+			}
+
+			if affCapture := s.lookUpAffinity(tableID); affCapture != "" {
+				log.Info("Dispatching table using affinity",
+					zap.String("cfID", s.cfID),
+					zap.Int("tableID", int(tableID)),
+					zap.String("target-capture", affCapture))
+				target = affCapture
+			}
+
+			replicaInfo := model.TableReplicaInfo{
+				StartTs:     task.CheckpointTs + 1,
+				MarkTableID: 0, // TODO support cyclic replication
+			}
+
+			log.Info("Dispatching table",
+				zap.Int64("table-id", tableID),
+				zap.String("target-capture", target),
+				zap.String("changefeed-id", s.cfID))
+
+			s.ownerState.DispatchTable(s.cfID, target, tableID, replicaInfo)
+			s.captureWorkloadDeltas[target]++
+		}
+	}
+
+	// handle deleting table
+	for tableID, captureID := range tableToCaptureMap {
+		if _, ok := pendingSet[tableID]; ok {
+			// Table has a pending deletion. Skip.
+			continue
+		}
+
+		if _, ok := tables[tableID]; !ok {
+			// Table should be deleted from the capture
+			log.Info("Stopping table",
+				zap.Int64("table-id", tableID),
+				zap.String("capture", captureID),
+				zap.String("changefeed-id", s.cfID))
+
+			s.ownerState.StartDeletingTable(s.cfID, captureID, tableID)
+			s.captureWorkloadDeltas[captureID]--
+		}
+	}
+
+	if s.needRebalance {
+		s.needRebalance = false
+		s.triggerRebalance()
+	}
+}
+
+// cleanUpOperations returns tablesIDs of tables that are NOT suitable for immediate redispatching.
+func (s *schedulerImpl) cleanUpOperations() []model.TableID {
+	var pendingList []model.TableID
+
+	for captureID, taskStatus := range s.ownerState.TaskStatuses[s.cfID] {
+		for tableID, operation := range taskStatus.Operation {
+			if operation.Status == model.OperFinished {
+				s.ownerState.CleanOperation(s.cfID, captureID, tableID)
+			} else {
+				if operation.Delete {
+					pendingList = append(pendingList, tableID)
+				}
+			}
+		}
+	}
+
+	return pendingList
+}
+
+// IsReady returns whether the scheduler is ready to process new requests.
+func (s *schedulerImpl) IsReady() bool {
+	return !s.cleanUpStaleCaptureStatus()
+}
+
+func (s *schedulerImpl) SetAffinity(tableID model.TableID, captureID model.CaptureID, ttl int) {
```

Comment:
Do we have document about how `affinity` works? What's the difference between `affinity` driven dispatcher and original workload based dispatching. Besides is the `affinity` memory only